layer1/executor/task_manager.go

```diff
@@ -4,282 +4,501 @@ import (
+	cancelChan     chan bool                      `json:"-"`
+	requestChan    <-chan managerRequest          `json:"-"`
+	responseChan   *executorResponseChan          `json:"-"`
+	logger         *logrus.Entry                  `json:"-"`
+	taskExecutor   *TaskExecutor                  `json:"-"`
 }
 
-// Creates a new TasksManager instance
-func NewTaskManager(txWatcher transaction.Watcher, database *db.Database, logger *logrus.Entry) (*TasksManager, error) {
-	taskManager := &TasksManager{
-		TxsBackup: make(map[string]*types.Transaction),
-		txWatcher: txWatcher,
-		database:  database,
-		logger:    logger,
+// newTaskManager creates a new TaskManager instance and recover the previous state from DB.
+func newTaskManager(mainCtx context.Context, eth layer1.Client, contracts layer1.AllSmartContracts, database *db.Database, logger *logrus.Entry, adminHandler monitorInterfaces.AdminHandler, requestChan <-chan managerRequest, txWatcher transaction.Watcher) (*TaskManager, error) {
+	taskManager := &TaskManager{
+		Schedule:     make(map[string]ManagerRequestInfo),
+		Responses:    make(map[string]ManagerResponseInfo),
+		mainCtx:      mainCtx,
+		database:     database,
+		eth:          eth,
+		contracts:    contracts,
+		adminHandler: adminHandler,
+		marshaller:   getTaskRegistry(),
+		cancelChan:   make(chan bool, 1),
+		requestChan:  requestChan,
+		responseChan: &executorResponseChan{erChan: make(chan ExecutorResponse, 100)},
+		logger:       logger,
 	}
 
-	err := taskManager.loadState()
+	err := taskManager.recoverState()
 	if err != nil {
 		taskManager.logger.Warnf("could not find previous State: %v", err)
 		if err != badger.ErrKeyNotFound {
 			return nil, err
 		}
 	}
 
+	tasksExecutor, err := newTaskExecutor(txWatcher, database, logger.WithField("Component", "TaskExecutor"))
+	if err != nil {
+		return nil, err
+	}
+	taskManager.taskExecutor = tasksExecutor
+
 	return taskManager, nil
 }
 
-func (tm *TasksManager) AddTxBackup(uuid string, tx *types.Transaction) error {
-	tm.Lock()
-	defer tm.Unlock()
-	tm.TxsBackup[uuid] = tx
-	return tm.persistState()
-}
+// start the infinite loop TaskManager execution.
+func (tm *TaskManager) start() {
+	tm.logger.Info("Starting task manager")
+	tm.logger.Info(strings.Repeat("-", 80))
+	tm.logger.Infof("Current Tasks: %d", len(tm.Schedule))
+	for id, task := range tm.Schedule {
+		tm.logger.Infof("...ID: %s Name: %s Between: %d and %d", id, task.Name, task.Start, task.End)
+	}
+	tm.logger.Info(strings.Repeat("-", 80))
 
-func (tm *TasksManager) RemoveTxBackup(uuid string) error {
-	tm.Lock()
-	defer tm.Unlock()
-	delete(tm.TxsBackup, uuid)
-	return tm.persistState()
+	go tm.eventLoop()
 }
 
-func (tm *TasksManager) GetTxBackup(uuid string) (*types.Transaction, bool) {
-	tm.RLock()
-	defer tm.RUnlock()
-	txn, present := tm.TxsBackup[uuid]
-	return txn, present
+// close the TaskManager execution.
+func (tm *TaskManager) close() {
+	tm.logger.Warn("Closing task manager")
+	tm.cancelChan <- true
```

Comment:
dangerous as this channel is closed by other code
may cause a panic
see line 107

Reply:
Good catch, ty. The line 107 is suppose to be triggered by `<-s.cancelChan` not by `<-tm.mainCtx.Done()`. Will change it. Making this modification, do you think it will be safe?

Reply:
Fixed